dnfdaemon-client/wrappers/dbus_package_wrapper.hpp

```diff
@@ -0,0 +1,91 @@
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef WRAPPERS_DBUS_PACKAGE_WRAPPER_HPP
+#define WRAPPERS_DBUS_PACKAGE_WRAPPER_HPP
+
+#include <string>
+#include <vector>
+
+namespace dnfdaemon::client {
+
+
+class DbusPackageWrapper {
+
+    // TODO (nsella) unify behaviour with DbusPackageKeyValue
+public:
+    DbusPackageWrapper(dnfdaemon::DbusTransactionItem transactionitem) {
+        ti = transactionitem;
+        full_nevra = std::get<1>(ti) + "-"
+            + std::get<3>(ti) + "-"
+            + std::get<4>(ti) + "."
+            + std::get<5>(ti);
+    }
+
+    uint64_t get_size() const { return std::get<7>(ti); }
+    std::string get_full_nevra() const { return full_nevra; }
+    std::string get_repo_id() const { return std::get<6>(ti); }
+
+private:
+    dnfdaemon::DbusTransactionItem ti;
+    std::string full_nevra;
+};
+
+
+class DbusPackageKeyValue {
+    // TODO rename class to DbusPackageWrapper
+public:
+    explicit DbusPackageKeyValue(const dnfdaemon::KeyValueMap & rawdata) : rawdata(rawdata){};
+    uint64_t get_id() const { return get_data(rawdata, "id"); }
+    int get_size() const { return get_data(rawdata, "size"); }
+    std::string get_name() const { return get_data(rawdata, "name"); }
+    std::string get_epoch() const { return get_data(rawdata, "epoch"); }
+    std::string get_version() const { return get_data(rawdata, "version"); }
+    std::string get_release() const { return get_data(rawdata, "release"); }
+    std::string get_arch() const { return get_data(rawdata, "arch"); }
+    std::string get_repo_id() const { return get_data(rawdata, "repo"); }
+    std::string get_nevra() const { return get_data(rawdata, "nevra"); }
+    std::string get_full_nevra() const { return get_data(rawdata, "full_nevra"); }
+    std::string get_sourcerpm() const { return get_data(rawdata, "sourcerpm"); }
+    std::string get_summary() const { return get_data(rawdata, "summary"); }
+    std::string get_url() const { return get_data(rawdata, "url"); }
+    std::string get_license() const { return get_data(rawdata, "license"); }
+    std::string get_description() const { return get_data(rawdata, "description"); }
+
+private:
+    template<typename... Args>
+    bool has_key(std::map<Args...> const& map, typename std::map<Args...>::key_type const& key) const {
+        return map.find(key) != map.end();
+    }
+
+    template <typename... Args>
+    typename std::map<Args...>::mapped_type get_data(std::map<Args...> const& map, typename std::map<Args...>::key_type const& key) const {
+        if(!has_key(map, key)) {
```

Comment:
`!map.contains(key)` can be used instead of `!has_key(map, key)`.
Ok `contains` requires >=C++20. But `map.count(key) == 0` can be used. So we don't need the own `has_key` private method.

But even that is not the good algorithm. Container is searched twice. First, if there is a `key` in it. And then again to get the value of the `key`.
We can do it in one search.
Something like this: 
```
auto it = map.find(key);
return it != map.end() ? it->second : typename std::map<Args...>::mapped_type();
```

Reply:
And why does the `get_data` method have a` map` input parameter and no direct access to `raw_data`? The method is now independent of any specific object of the `DbusPackageKeyValue` class.
If this is intended, please mark the method as `static`. It's a private method, so your intent may change in the future.